### ISSUE TITLE
Priority inversion issue on accessing channel from different threads

### ISSUE THREAD

**Issue by AndrewSB** [GetStream/stream-chat-swift#2355]
hey! we ran into an issue that caused main thread lock for 10+ seconds in some cases. here's what the thread state looks like during that

Thread 44 is running some sort of long running operation

<img width="1512" alt="Снимок экрана 2022-10-27 в 22 12 10" src="https://user-images.githubusercontent.com/3814772/198925875-f1ed8471-c5b1-49e7-8e86-181be2cceb7b.png">

Thread 1, then wants to read something 

<img width="1512" alt="Снимок экрана 2022-10-27 в 22 09 01" src="https://user-images.githubusercontent.com/3814772/198925995-4e6b9c0c-c3d5-4471-af6b-3535dee66f7e.png">

The implementation of that read on thread 1 calls into this `reset()` function. which is mutating an `@Atomic` property

<img width="1512" alt="Снимок экрана 2022-10-27 в 22 11 05" src="https://user-images.githubusercontent.com/3814772/198926069-6bfc0362-8836-4904-ae59-be6486be6a4f.png">

Since Thread 44 is still running this long running operation which has is scheduled on that `@Atomic` serial queue, Thread 1 needs to wait for that operation on a lower priority thread to finish before it can complete and unblock UI

**Comment by AndrewSB** [comment]
we've worked around this by moving what used to be on the main thread, off the main thread. it doesn't _feel_ right though, because the main thread call we were making into Stream is a view related function, so i'd expect to call that on the main thread.

is that object not thread safe?

**Comment by AndrewSB** [comment]
we've worked around this by moving what used to be on the main thread, off the main thread. it doesn't _feel_ right though, because the main thread call we were making into Stream is a view related function, so i'd expect to call that on the main thread.

is that object not thread safe?

**Comment by polqf** [comment]
Hi @AndrewSB ,

I am curious to understand what is it that supposedly caused a thread to hang for 10+ seconds.
Based on what you shared, it seems like a DB read was performed and that's what caused other threads to wait. It is unlikely that a DB read takes 10s, so would be nice to understand what you were doing.
We use locking mechanisms to ensure atomicity for certain pieces of our codebase, but we have not seen these issues in our experience. 

That said, it is discouraged to use our SDK APIs outside of the main thread. We are working on ensuring that our non UI APIs are fully safe to be used outside of the main thread, but as of now there are some edge cases.

**Comment by polqf** [comment]
@AndrewSB given there are no further updates on this, we'll be closing it for now

**Comment by jfru** [comment]
we've found one case of this when we fetch the `activeMembers` of a channel on a background thread. we've worked around that by dispatching to the main thread before we make that function call.

do you think you and i can get on a quick call where i show you the deadlock & you see if its possible for you to fix upstream?

**Comment by nuno-vieira** [comment]
Hi @jfru,

Can you share the code snippet of how you are doing this?

Best,
Nuno
